lib/controllers/admin_services.ts

```diff
@@ -0,0 +1,217 @@
+/*
+ * Implements the API handlers for the Services resource.
+ */
+import * as express from "express";
+
+import { IRetrievedService, IService, ServiceModel } from "../models/service";
+
+import {
+  AzureApiAuthMiddleware,
+  IAzureApiAuthorization,
+  UserGroup
+} from "../utils/middlewares/azure_api_auth";
+
+import {
+  IRequestMiddleware,
+  withRequestMiddlewares,
+  wrapRequestHandler
+} from "../utils/request_middleware";
+
+import { PathParamMiddleware } from "../utils/middlewares/path_param";
+
+import {
+  IResponseErrorInternal,
+  IResponseErrorQuery,
+  IResponseErrorValidation,
+  IResponseSuccessJson,
+  ResponseErrorQuery,
+  ResponseErrorValidation,
+  ResponseSuccessJson
+} from "../utils/response";
+
+import { left, right } from "../utils/either";
+import {
+  isNonEmptyString,
+  NonEmptyString,
+  ObjectIdGenerator,
+  ulidGenerator
+} from "../utils/strings";
+
+type ICreateServiceHandler = (
+  auth: IAzureApiAuthorization,
+  service: IService
+) => Promise<
+  | IResponseSuccessJson<IRetrievedService>
+  | IResponseErrorValidation
+  | IResponseErrorQuery
+  | IResponseErrorInternal
+>;
+
+type IUpdateServiceHandler = (
+  auth: IAzureApiAuthorization,
+  serviceId: NonEmptyString,
+  service: IRetrievedService
```

Comment:
this should probably be a `INewService` ?

Reply:
why new if updating ?

Reply:
the fact that `New` is in the name is misleading, think about that type as`IServiceWithoutId` 😄 

the problem in using `IRetrievedService` here is that it extends `DocumentDb.RetrievedDocument` and `IVersionedModel`... these interfaces contain attributes that should not be a concern of the client